### PR TITLE
Fixes #385

### LINKED ISSUE #385: cfunction issues
Couldn't really find a MWE for this, but if I `include` [this](https://github.com/SciML/Sundials.jl/blob/master/test/common_interface/mass_matrix.jl) file (after adding Sundials, of course) and call 
```
using JuliaInterpreter
@interpret solve(prob, ARKODE(),abstol=1e-8,reltol=1e-8)
```
I'm getting
```
MethodError: no method matching iterate(::Type{Int32})
Closest candidates are:
  iterate(!Matched::Core.SimpleVector) at essentials.jl:603
  iterate(!Matched::Core.SimpleVector, !Matched::Any) at essentials.jl:603
  iterate(!Matched::ExponentialBackOff) at error.jl:253
  ...
iterate at generator.jl:44 [inlined]
collect(::Base.Generator{DataType,JuliaInterpreter.var"#98#99"{JuliaInterpreter.FrameData,UnionAll}}) at array.jl:665
map(::Function, ::Type{T} where T) at abstractarray.jl:2098
evaluate_foreigncall(::Frame, ::Expr) at interpret.jl:165
eval_rhs(::Any, ::Frame, ::Expr) at interpret.jl:389
step_expr!(::Any, ::Frame, ::Any, ::Bool) at interpret.jl:527
step_expr! at interpret.jl:566 [inlined]
finish!(::Any, ::Frame, ::Bool) at commands.jl:14
finish_and_return! at commands.jl:29 [inlined]
evaluate_call_recurse!(::Any, ::Frame, ::Expr; enter_generated::Bool) at interpret.jl:240
evaluate_call_recurse! at interpret.jl:202 [inlined]
eval_rhs(::Any, ::Frame, ::Expr) at interpret.jl:387
step_expr!(::Any, ::Frame, ::Any, ::Bool) at interpret.jl:443
step_expr! at interpret.jl:566 [inlined]
finish!(::Any, ::Frame, ::Bool) at commands.jl:14
finish_and_return! at commands.jl:29 [inlined]
...
```

The stacktrace points to [these lines](https://github.com/JuliaDebug/JuliaInterpreter.jl/blob/00fe0b1bfeb6282acf79e3249db453faafcc4a1a/src/interpret.jl#L163-L165), which apparently can't handle one of the `@cfunction`s in Sundials.jl.

### PR DESCRIPTION
**Pull request by goerch**
Tested with both cases from the issue.

### CODE REVIEW

**Comment by aviatesk** [comment]
Can you add a test case? And CI seems to be broken because of this change.

**Comment by goerch** [comment]
> Can you add a test case? 

I have no better reproducer than the two examples in #385. Which one do we want? Both?

> And CI seems to be broken because of this change.

Interesting, the change does not work on 1.6.

File: src/interpret.jl [hunk 1]
```diff
@@ -157,11 +157,25 @@ function evaluate_foreigncall(frame::Frame, call_expr::Expr)
     scope = frame.framecode.scope
     data = frame.framedata
     if !isempty(data.sparams) && scope isa Method
-        sig = scope.sig
+        sig = scope.sig
```
Comment (timholy) [inline src/interpret.jl:160]: Whitespace

File: src/interpret.jl [hunk 2]
```diff
@@ -157,11 +157,25 @@ function evaluate_foreigncall(frame::Frame, call_expr::Expr)
...
+                args[4] = Core.svec(map(args[4]) do arg
+                    instantiate_type_in_env(arg, sig, data.sparams)
+                end...)
+            end
+        else
+            args[3] = instantiate_type_in_env(args[3], sig, data.sparams)
+            args[4] = Core.svec(map(args[4]) do arg
+                instantiate_type_in_env(arg, sig, data.sparams)
+            end...)
+        end
```
Comment (timholy) [inline src/interpret.jl:178]: How much of the argument-position issue is specific to `:cfunction`? We may just be off by one index in `args` for `cfunction`.

**Review by timholy — COMMENTED** [review]
This looks great, thanks for tackling this @goerch!

File: src/interpret.jl (hunk 2 above)
Comment (goerch) [reply src/interpret.jl:178]: Will check.

File: src/interpret.jl (hunk 1 above)
Comment (goerch) [reply src/interpret.jl:160]: Thanks!

File: src/interpret.jl [hunk 3]
```diff
@@ -163,11 +163,25 @@ function evaluate_foreigncall(@nospecialize(recurse), frame::Frame, call_expr::E
...
```
Comment (aviatesk) [inline src/interpret.jl:184]:
```suggestion
        @static if VERSION < v"1.7.0"
            arg3 = args[3]
            if arg3 isa Core.SimpleVector
                args[3] = Core.svec(map(arg3) do arg
                    instantiate_type_in_env(arg, sig, data.sparams)
                end...)
            else
                args[3] = instantiate_type_in_env(arg3, sig, data.sparams)
                args[4] = Core.svec(map(args[4]::Core.SimpleVector) do arg
                    instantiate_type_in_env(arg, sig, data.sparams)
                end...)
            end
        else
            args[3] = instantiate_type_in_env(args[3], sig, data.sparams)
            args[4] = Core.svec(map(args[4]::Core.SimpleVector) do arg
                instantiate_type_in_env(arg, sig, data.sparams)
            end...)
        end
```
for better inferrability.

File: src/interpret.jl (hunk 1 above)
Comment (aviatesk) [reply src/interpret.jl:166]: Please remove these ws.

File: src/interpret.jl (hunk 2 above)
Comment (goerch) [reply src/interpret.jl:185]: I searched for documentation about `:cfunction` in AST and found nothing ;( So i added some logging to test and see on 1.6.4

```
head = :foreigncall
args[1] = :(Core.tuple("cholmod_l_allocate_sparse", :libcholmod))
args[2] = Ptr{SuiteSparse.CHOLMOD.C_Sparse{Float64}}
args[3] = svec(UInt64, UInt64, UInt64, Int32, Int32, Int32, Int32, Ptr{Nothing})
head = :foreigncall
args[1] = :(Core.tuple("cholmod_l_check_sparse", :libcholmod))
args[2] = Int32
args[3] = svec(Ptr{SuiteSparse.CHOLMOD.C_Sparse{Tv<:Union{Float64, ComplexF64}}}, Ptr{UInt8})
head = :foreigncall
args[1] = :(Core.tuple("cholmod_l_analyze", :libcholmod))
args[2] = Ptr{SuiteSparse.CHOLMOD.C_Factor{Float64}}
args[3] = svec(Ptr{SuiteSparse.CHOLMOD.C_Sparse{Tv<:Union{Float64, ComplexF64}}}, Ptr{UInt8})
head = :foreigncall
args[1] = :(Core.tuple("cholmod_l_factorize_p", :libcholmod))
args[2] = Int32
args[3] = svec(Ptr{SuiteSparse.CHOLMOD.C_Sparse{Tv<:Union{Float64, ComplexF64}}}, Ref{ComplexF64}, Ptr{Int64}, UInt64, Ptr{SuiteSparse.CHOLMOD.C_Factor{Tv<:Union{Float64, ComplexF64}}}, Ptr{UInt8})
head = :foreigncall
args[1] = :(Core.tuple("cholmod_l_allocate_dense", :libcholmod))
args[2] = Ptr{SuiteSparse.CHOLMOD.C_Dense{Float64}}
args[3] = svec(UInt64, UInt64, UInt64, Int32, Ptr{Nothing})
head = :foreigncall
args[1] = :(Core.tuple("cholmod_l_solve", :libcholmod))
args[2] = Ptr{SuiteSparse.CHOLMOD.C_Dense{Float64}}
args[3] = svec(Int32, Ptr{SuiteSparse.CHOLMOD.C_Factor{Tv<:Union{Float64, ComplexF64}}}, Ptr{SuiteSparse.CHOLMOD.C_Dense{Tv<:Union{Float64, ComplexF64}}}, Ptr{UInt8})
head = :cfunction
args[1] = Ptr{Nothing}
args[2] = :($(QuoteNode(FunctionWrappers.CallWrapper{Int64}())))
args[3] = Int64
args[4] = svec(Ref{var"#35#36"})```
```

and on 1.7.0

```
head = :cfunction
args = Any[Ptr{Nothing}, :($(QuoteNode(FunctionWrappers.CallWrapper{Int64}()))), Int64, svec(Ref{var"#35#36"}), :(:ccall)]
```

Oops.

Comment (goerch) [reply src/interpret.jl:185]: More logging:

1.6.4:

```
head = :foreigncall
args[1] = :(:jl_sizeof_jl_options)
args[2] = Int64
args[3] = svec()
args[4] = 0
data.sparams = Any[]
scope = Base
head = :foreigncall
args[1] = :cf
args[2] = Int64
args[3] = svec(Int64, Int64)
args[4] = 0
data.sparams = Any[]
scope = Main
head = :foreigncall
args[1] = Ptr{Nothing} @0x000000005a8d5f90
args[2] = Int64
args[3] = svec(Int64, Int64)
args[4] = 0
data.sparams = Any[]
scope = Main
head = :cfunction
args[1] = Ptr{Nothing}
args[2] = :($(QuoteNode(fcfun)))
args[3] = Int64
args[4] = svec(Int64, Int64)
data.sparams = Any[]
scope = cfcfun() in Main at C:\Users\Win10\Documents\GitHub\JuliaInterpreter.jl\test\interpret.jl:162
head = :foreigncall
args[1] = Ptr{Nothing} @0x000000005a8d9630
args[2] = Int64
args[3] = svec(UInt8, Int64)
args[4] = 0
data.sparams = Any[]
scope = Main
head = :foreigncall
args[1] = :(Core.tuple("cholmod_l_allocate_sparse", :libcholmod))
args[2] = Ptr{SuiteSparse.CHOLMOD.C_Sparse{Tv<:Union{Float64, ComplexF64}}}
args[3] = svec(UInt64, UInt64, UInt64, Int32, Int32, Int32, Int32, Ptr{Nothing})
args[4] = 0
data.sparams = Any[Float64]
scope = allocate_sparse(nrow::Integer, ncol::Integer, nzmax::Integer, sorted::Bool, packed::Bool, stype::Integer, ::Type{Tv}) where Tv<:Union{Float64, ComplexF64} in SuiteSparse.CHOLMOD at C:\Users\Win10\AppData\Local\Programs\Julia-1.6.4\share\julia\stdlib\v1.6\SuiteSparse\src\cholmod.jl:483
head = :foreigncall
args[1] = :(Core.tuple("cholmod_l_check_sparse", :libcholmod))
args[2] = Int32
args[3] = svec(Ptr{SuiteSparse.CHOLMOD.C_Sparse{Tv<:Union{Float64, ComplexF64}}}, Ptr{UInt8})
args[4] = 0
data.sparams = Any[Float64]
scope = check_sparse(A::SuiteSparse.CHOLMOD.Sparse{Tv}) where Tv<:Union{Float64, ComplexF64} in SuiteSparse.CHOLMOD at C:\Users\Win10\AppData\Local\Programs\Julia-1.6.4\share\julia\stdlib\v1.6\SuiteSparse\src\cholmod.jl:542
head = :foreigncall
args[1] = :(Core.tuple("cholmod_l_analyze", :libcholmod))
args[2] = Ptr{SuiteSparse.CHOLMOD.C_Factor{Tv<:Union{Float64, ComplexF64}}}
args[3] = svec(Ptr{SuiteSparse.CHOLMOD.C_Sparse{Tv<:Union{Float64, ComplexF64}}}, Ptr{UInt8})
args[4] = 0
data.sparams = Any[Float64]
scope = analyze(A::SuiteSparse.CHOLMOD.Sparse{Tv}, cmmn::Vector{UInt8}) where Tv<:Union{Float64, ComplexF64} in SuiteSparse.CHOLMOD at C:\Users\Win10\AppData\Local\Programs\Julia-1.6.4\share\julia\stdlib\v1.6\SuiteSparse\src\cholmod.jl:718
head = :foreigncall
args[1] = :(Core.tuple("cholmod_l_factorize_p", :libcholmod))
args[2] = Int32
args[3] = svec(Ptr{SuiteSparse.CHOLMOD.C_Sparse{Tv<:Union{Float64, ComplexF64}}}, Ref{ComplexF64}, Ptr{Int64}, UInt64, Ptr{SuiteSparse.CHOLMOD.C_Factor{Tv<:Union{Float64, ComplexF64}}}, Ptr{UInt8})
args[4] = 0
data.sparams = Any[Float64]
scope = factorize_p!(A::SuiteSparse.CHOLMOD.Sparse{Tv}, β::Real, F::SuiteSparse.CHOLMOD.Factor{Tv}, cmmn::Vector{UInt8}) where Tv<:Union{Float64, ComplexF64} in SuiteSparse.CHOLMOD at C:\Users\Win10\AppData\Local\Programs\Julia-1.6.4\share\julia\stdlib\v1.6\SuiteSparse\src\cholmod.jl:739
head = :foreigncall
args[1] = :(Core.tuple("cholmod_l_allocate_dense", :libcholmod))
args[2] = Ptr{SuiteSparse.CHOLMOD.C_Dense{Tv<:Union{Float64, ComplexF64}}}
args[3] = svec(UInt64, UInt64, UInt64, Int32, Ptr{Nothing})
args[4] = 0
data.sparams = Any[Float64]
scope = allocate_dense(m::Integer, n::Integer, d::Integer, ::Type{Tv}) where Tv<:Union{Float64, ComplexF64} in SuiteSparse.CHOLMOD at C:\Users\Win10\AppData\Local\Programs\Julia-1.6.4\share\julia\stdlib\v1.6\SuiteSparse\src\cholmod.jl:413
head = :foreigncall
args[1] = :(Core.tuple("cholmod_l_solve", :libcholmod))
args[2] = Ptr{SuiteSparse.CHOLMOD.C_Dense{Tv<:Union{Float64, ComplexF64}}}
args[3] = svec(Int32, Ptr{SuiteSparse.CHOLMOD.C_Factor{Tv<:Union{Float64, ComplexF64}}}, Ptr{SuiteSparse.CHOLMOD.C_Dense{Tv<:Union{Float64, ComplexF64}}}, Ptr{UInt8})
args[4] = 0
data.sparams = Any[Float64]
scope = solve(sys::Integer, F::SuiteSparse.CHOLMOD.Factor{Tv}, B::SuiteSparse.CHOLMOD.Dense{Tv}) where Tv<:Union{Float64, ComplexF64} in SuiteSparse.CHOLMOD at C:\Users\Win10\AppData\Local\Programs\Julia-1.6.4\share\julia\stdlib\v1.6\SuiteSparse\src\cholmod.jl:749
head = :cfunction
args[1] = Ptr{Nothing}
args[2] = :($(QuoteNode(+)))
args[3] = Int64
args[4] = svec(Int64, Int64)
data.sparams = Any[]
scope = (::var"#f#34")() in Main at C:\Users\Win10\Documents\GitHub\JuliaInterpreter.jl\test\interpret.jl:786
head = :cfunction
args[1] = Ptr{Nothing}
args[2] = :($(QuoteNode(FunctionWrappers.CallWrapper{Int64}())))
args[3] = Int64
args[4] = svec(Ref{var"#35#36"})
data.sparams = Any[Int64, Tuple{}, var"#35#36"]
scope = gen_fptr(::Type{Ret}, ::Type{Args}, ::Type{objT}) where {Ret, Args, objT} in FunctionWrappers at C:\Users\Win10\.julia\packages\FunctionWrappers\8xdVB\src\FunctionWrappers.jl:81
head = :foreigncall
args[1] = :(:strlen)
args[2] = UInt64
args[3] = svec(Cstring)
args[4] = 0
data.sparams = Any[]
scope = Main.anonymous
head = :foreigncall
args[1] = :(:asprintf)
args[2] = Int32
args[3] = svec(Ptr{Ptr{Int8}}, Cstring, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, Float32, Float32, Float32, Float32, Float32, Float32, Float32, Float32, Float32)
args[4] = 2
data.sparams = Any[]
scope = Main.anonymous
head = :foreigncall
args[1] = :(:free)
args[2] = Nothing
args[3] = svec(Cstring)
args[4] = 0
data.sparams = Any[]
scope = Main.anonymous
```

On 1.7.0:

```
head = :foreigncall
args[1] = :(:jl_sizeof_jl_options)
args[2] = Int64
args[3] = svec()
args[4] = 0
data.sparams = Any[]
scope = Base
head = :foreigncall
args[1] = :cf
args[2] = Int64
args[3] = svec(Int64, Int64)
args[4] = 0
data.sparams = Any[]
scope = Main
head = :foreigncall
args[1] = Ptr{Nothing} @0x0000000062462e00
args[2] = Int64
args[3] = svec(Int64, Int64)
args[4] = 0
data.sparams = Any[]
scope = Main
head = :cfunction
args[1] = Ptr{Nothing}
args[2] = :($(QuoteNode(fcfun)))
args[3] = Int64
args[4] = svec(Int64, Int64)
data.sparams = Any[]
scope = cfcfun() in Main at C:\Users\Win10\Documents\GitHub\JuliaInterpreter.jl\test\interpret.jl:162
head = :foreigncall
args[1] = Ptr{Nothing} @0x00000000624665c0
args[2] = Int64
args[3] = svec(UInt8, Int64)
args[4] = 0
data.sparams = Any[]
scope = Main
head = :cfunction
args[1] = Ptr{Nothing}
args[2] = :($(QuoteNode(+)))
args[3] = Int64
args[4] = svec(Int64, Int64)
data.sparams = Any[]
scope = (::var"#f#34")() in Main at C:\Users\Win10\Documents\GitHub\JuliaInterpreter.jl\test\interpret.jl:786
head = :cfunction
args[1] = Ptr{Nothing}
args[2] = :($(QuoteNode(FunctionWrappers.CallWrapper{Int64}())))
args[3] = Int64
args[4] = svec(Ref{var"#35#36"})
data.sparams = Any[Int64, Tuple{}, var"#35#36"]
scope = gen_fptr(::Type{Ret}, ::Type{Args}, ::Type{objT}) where {Ret, Args, objT} in FunctionWrappers at C:\Users\Win10\.julia\packages\FunctionWrappers\8xdVB\src\FunctionWrappers.jl:81
head = :foreigncall
args[1] = :(:strlen)
args[2] = UInt64
args[3] = svec(Cstring)
args[4] = 0
data.sparams = Any[]
scope = Main.anonymous
head = :foreigncall
args[1] = :(:asprintf)
args[2] = Int32
args[3] = svec(Ptr{Ptr{Int8}}, Cstring, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, Float32, Float32, Float32, Float32, Float32, Float32, Float32, Float32, Float32)
args[4] = 2
data.sparams = Any[]
scope = Main.anonymous
head = :foreigncall
args[1] = :(:free)
args[2] = Nothing
args[3] = svec(Cstring)
args[4] = 0
data.sparams = Any[]
scope = Main.anonymous
```

**Comment by goerch** [comment]
I have checked the call sites of the Julia test suite with https://github.com/goerch/JuliaInterp.jl and see something like

```
:cfunction = :cfunction
args = Any[Ptr{Nothing}, :(:identity), :(Core.apply_type(VecElement, Float64)), :(Core.svec(Core.apply_type(VecElement, Float64))), :(:ccall)]
:cfunction = :cfunction
args = Any[Ptr{Nothing}, :(:identity), :(Core.apply_type(VecElement, Core.apply_type(NTuple, 2, Int8))), :(Core.svec(Core.apply_type(VecElement, Core.apply_type(NTuple, 2, Int8)))), :(:ccall)]
:cfunction = :cfunction
args = Any[Ptr{Nothing}, :(:foo13031), :Cint, :(Core.svec(Core.apply_type(Ref, Core.apply_type(Tuple)))), :(:ccall)]
:cfunction = :cfunction
args = Any[Ptr{Nothing}, :(:foo13031), :Cint, :(Core.svec(Core.apply_type(Ref, Core.apply_type(Tuple)), Core.apply_type(Ref, Core.apply_type(Tuple)), Cint)), :(:ccall)]
:cfunction = :cfunction
args = Any[Ptr{Nothing}, :(:unstable26078), :Int32, :(Core.svec(Int32)), :(:ccall)]
:cfunction = :cfunction
args = Any[Ptr{Nothing}, :($(QuoteNode(:(Base.last)))), :String, :(Core.svec(Core.apply_type(Tuple, Int, String))), :(:ccall)]
:cfunction = :cfunction
args = Any[Ptr{Nothing}, :($(QuoteNode(:($(Expr(Symbol("hygienic-scope"), :(function (fptr::Ptr{Cvoid}, args_ptr::Ptr{Cvoid}, retval_ptr::Ptr{Cvoid})
    #= threadcall.jl:33 =#
    #= threadcall.jl:34 =#
    p = args_ptr
    arg1 = unsafe_load(convert(Ptr{$(Expr(:escape, :Csize_t))}, p))
    p += Core.sizeof($(Expr(:escape, :Csize_t)))
    ret = ccall(fptr, $(Expr(:escape, :(Ptr{Cint}))), ($(Expr(:escape, :Csize_t)),), arg1)
    unsafe_store!(convert(Ptr{$(Expr(:escape, :(Ptr{Cint})))}, retval_ptr), ret)
    return Int(Core.sizeof($(Expr(:escape, :(Ptr{Cint})))))
    end), Base)))))), :(Base.Int), :(Core.svec(Core.apply_type(Base.Ptr, Base.Cvoid), Core.apply_type(Base.Ptr, Base.Cvoid), Core.apply_type(Base.Ptr, Base.Cvoid))), :(:ccall)]
[and so on]
```

So it looks to me like the patch is reasonable?

**Comment by KristofferC** [comment]
> So it looks to me like the patch is reasonable?

I haven't looked into detail but since it obviously makes the tests better I think this is good to go. Any further thoughts @timholy?

**Pull request merged**